ecdsa/src/hazmat.rs

```diff
@@ -216,8 +217,35 @@ where
 #[cfg_attr(docsrs, doc(cfg(feature = "digest")))]
 pub trait DigestPrimitive: PrimeCurve {
     /// Preferred digest to use when computing ECDSA signatures for this
-    /// elliptic curve. This should be a member of the SHA-2 family.
+    /// elliptic curve. This is typically a member of the SHA-2 family.
+    // TODO(tarcieri): add BlockSizeUser + FixedOutput(Reset) bounds in next breaking release
+    // These bounds ensure the digest algorithm can be used for HMAC-DRBG for RFC6979
     type Digest: Digest;
+
+    /// Compute field bytes for a prehash (message digest), either zero-padding
+    /// or truncating if the prehash size does not match the field size.
+    fn prehash_to_field_bytes(prehash: &[u8]) -> Result<FieldBytes<Self>> {
+        // Minimum allowed prehash size is half the field size
+        if prehash.len() < Self::UInt::BYTE_SIZE / 2 {
+            return Err(Error::new());
+        }
+
+        let mut field_bytes = FieldBytes::<Self>::default();
+
+        match prehash.len().cmp(&Self::UInt::BYTE_SIZE) {
+            cmp::Ordering::Equal => field_bytes.copy_from_slice(prehash),
+            cmp::Ordering::Less => {
+                // If prehash is smaller than the field size, pad with zeroes
+                field_bytes[..prehash.len()].copy_from_slice(prehash);
```

Comment:
I'm not confident enough but if I read RFC correctly I believe this should be like:

```rust
field_bytes[(field_bytes.len()-prehash.len())..].copy_from_slice(prehash);
```

https://datatracker.ietf.org/doc/html/rfc6979#section-2.3.2

> otherwise, qlen-blen bits (of value zero) are added to the left of the sequence

the present implementation adds zero on the right so the prehash gets "left shift"'d, while informal description in RFC says:

> in many  software implementations of arithmetics on big integers, that  division is equivalent to a "right shift" by blen-qlen bits.

Reply:
Opened a PR for this

➡️  https://github.com/RustCrypto/signatures/pull/547